pkg/reconcilers/deployment.go

```diff
@@ -0,0 +1,272 @@
+}
+
+// DeploymentRemoveDuplicateEnvVarMutator ensures pod env vars are not duplicated
+func DeploymentRemoveDuplicateEnvVarMutator(_, existing *k8sappsv1.Deployment) (bool, error) {
+	updated := false
+	for idx := range existing.Spec.Template.Spec.Containers {
+		prunedEnvs := helper.RemoveDuplicateEnvVars(existing.Spec.Template.Spec.Containers[idx].Env)
+		if !reflect.DeepEqual(existing.Spec.Template.Spec.Containers[idx].Env, prunedEnvs) {
+			existing.Spec.Template.Spec.Containers[idx].Env = prunedEnvs
+			updated = true
+		}
+	}
+
+	return updated, nil
+}
+
+// DeploymentPriorityClassMutator ensures priorityclass is reconciled
+func DeploymentPriorityClassMutator(desired, existing *k8sappsv1.Deployment) (bool, error) {
+	updated := false
+
+	if existing.Spec.Template.Spec.PriorityClassName != desired.Spec.Template.Spec.PriorityClassName {
+		existing.Spec.Template.Spec.PriorityClassName = desired.Spec.Template.Spec.PriorityClassName
+		updated = true
+	}
+
+	return updated, nil
+}
+
+// DeploymentStrategyMutator ensures desired strategy
+func DeploymentStrategyMutator(desired, existing *k8sappsv1.Deployment) (bool, error) {
+	updated := false
+
+	if !reflect.DeepEqual(existing.Spec.Strategy, desired.Spec.Strategy) {
+		existing.Spec.Strategy = desired.Spec.Strategy
+		updated = true
+	}
+
+	return updated, nil
+}
+
+// DeploymentTopologySpreadConstraintsMutator ensures TopologySpreadConstraints is reconciled
+func DeploymentTopologySpreadConstraintsMutator(desired, existing *k8sappsv1.Deployment) (bool, error) {
+	updated := false
+
+	if !reflect.DeepEqual(existing.Spec.Template.Spec.TopologySpreadConstraints, desired.Spec.Template.Spec.TopologySpreadConstraints) {
+		diff := cmp.Diff(existing.Spec.Template.Spec.TopologySpreadConstraints, desired.Spec.Template.Spec.TopologySpreadConstraints)
+		log.Info(fmt.Sprintf("%s spec.template.spec.TopologySpreadConstraints has changed: %s", common.ObjectInfo(desired), diff))
+		existing.Spec.Template.Spec.TopologySpreadConstraints = desired.Spec.Template.Spec.TopologySpreadConstraints
+		updated = true
+	}
+
+	return updated, nil
+}
+
+// DeploymentPodTemplateAnnotationsMutator ensures Pod Template Annotations is reconciled
+func DeploymentPodTemplateAnnotationsMutator(desired, existing *k8sappsv1.Deployment) (bool, error) {
+	updated := false
+
+	helper.MergeMapStringString(&updated, &existing.Spec.Template.Annotations, desired.Spec.Template.Annotations)
+
+	return updated, nil
+}
+
+// CreateImageTriggerAnnotationString creates the annotation for Deployments to leverage existing ImageStreamTags
+func CreateImageTriggerAnnotationString(containers map[string]string) string {
+	// Need to sort the map to guarantee the range iteration order, otherwise the operator may unnecessarily update the annotations.
+	var keys []string
+	for k := range containers {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	annotationString := ""
+	if len(containers) > 0 {
+		annotationString += `[`
```

Comment:
Even though it is correct, what about this alternative implementation (feel free to customize it as you wish)

```diff
diff --git a/pkg/3scale/amp/component/apicast.go b/pkg/3scale/amp/component/apicast.go
index 9b8a102c..9dcd90fd 100644
--- a/pkg/3scale/amp/component/apicast.go
+++ b/pkg/3scale/amp/component/apicast.go
@@ -676,9 +676,15 @@ func (apicast *Apicast) productionDeploymentAnnotations() map[string]string {
 	annotations := map[string]string{}
 
 	// Image trigger annotation should always be present
-	imageTriggerString := reconcilers.CreateImageTriggerAnnotationString(map[string]string{
-		ApicastProductionInitContainerName: fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
-		ApicastProductionName:              fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+	imageTriggerString := reconcilers.CreateImageTriggerAnnotationString([]reconcilers.ContainerImage{
+		{
+			Name: ApicastProductionInitContainerName,
+			Tag:  fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+		},
+		{
+			Name: ApicastProductionName,
+			Tag:  fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+		},
 	})
 	annotations[reconcilers.DeploymentImageTriggerAnnotation] = imageTriggerString
 
@@ -702,8 +708,11 @@ func (apicast *Apicast) stagingDeploymentAnnotations() map[string]string {
 	annotations := map[string]string{}
 
 	// Image trigger annotation should always be present
-	imageTriggerString := reconcilers.CreateImageTriggerAnnotationString(map[string]string{
-		ApicastStagingName: fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+	imageTriggerString := reconcilers.CreateImageTriggerAnnotationString([]reconcilers.ContainerImage{
+		{
+			Name: ApicastStagingName,
+			Tag:  fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+		},
 	})
 	annotations[reconcilers.DeploymentImageTriggerAnnotation] = imageTriggerString
 
diff --git a/pkg/reconcilers/deployment.go b/pkg/reconcilers/deployment.go
index eb3f459d..b8dc3ccc 100644
--- a/pkg/reconcilers/deployment.go
+++ b/pkg/reconcilers/deployment.go
@@ -1,10 +1,9 @@
 package reconcilers
 
 import (
+	"encoding/json"
 	"fmt"
 	"reflect"
-	"sort"
-	"strings"
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
@@ -249,24 +248,44 @@ func DeploymentPodTemplateAnnotationsMutator(desired, existing *k8sappsv1.Deploy
 	return updated, nil
 }
 
+type ContainerImage struct {
+	Name string
+	Tag  string
+}
+
+type ImageTriggerFrom struct {
+	Kind string `json:"kind"`
+	Name string `json:"name"`
+	// +optional
+	Namespace *string `json:"namespace,omitempty"`
+}
+
+type ImageTrigger struct {
+	From      ImageTriggerFrom `json:"from"`
+	FieldPath string           `json:"fieldPath"`
+	// +optional
+	Pused bool `json:"paused,omitempty"`
+}
+
 // CreateImageTriggerAnnotationString creates the annotation for Deployments to leverage existing ImageStreamTags
-func CreateImageTriggerAnnotationString(containers map[string]string) string {
-	// Need to sort the map to guarantee the range iteration order, otherwise the operator may unnecessarily update the annotations.
-	var keys []string
-	for k := range containers {
-		keys = append(keys, k)
+func CreateImageTriggerAnnotationString(containers []ContainerImage) string {
+
+	triggers := make([]ImageTrigger, len(containers))
+
+	for _, container := range containers {
+		triggers = append(triggers, ImageTrigger{
+			From: ImageTriggerFrom{
+				Kind: "ImageStreamTag",
+				Name: container.Tag,
+			},
+			FieldPath: fmt.Sprintf(`spec.template.spec.containers[?(@.name=="%v")].image`, container.Name),
+		})
 	}
-	sort.Strings(keys)
 
-	annotationString := ""
-	if len(containers) > 0 {
-		annotationString += `[`
-		for _, c := range keys {
-			annotationString += fmt.Sprintf(`{"from":{"kind":"ImageStreamTag","name":"%v"},"fieldPath":"spec.template.spec.containers[?(@.name==\"%v\")].image"},`, containers[c], c)
-		}
-		annotationString = strings.TrimSuffix(annotationString, ",")
-		annotationString += `]`
+	if len(triggers) > 0 {
+		jsonData, _ := json.Marshal(triggers)
+		return string(jsonData)
 	}
 
-	return annotationString
+	return ""
 }
```

Reply:
Thank you for this, this is a much more elegant solution. I just tested and pushed the changes.